cmd/api/middleware.go

```diff
@@ -112,11 +111,21 @@ func (app *application) recoverPanic(next http.Handler) http.Handler {
 	})
 }
 
-func checkPermissions(user *models.User, permissions []string) bool {
+func checkPermissions(user *models.User, permissions []string, reqUUID string) bool {
+	ok := false
 	for _, userPermission := range user.Permissions {
 		for _, requiredPermission := range permissions {
 			if userPermission.Permission.Name == requiredPermission {
-				return true
+				if requiredPermission == "self:read" || requiredPermission == "self:write" {
+					if user.UUID.String() == reqUUID {
+						ok = true
+					}
+				}
+
+				if requiredPermission == "user:read" || requiredPermission == "user:write" {
```

Comment:
Correct me if I am wrong, but it seems like `"user:read/user:write"` would be a permission given to an admin and would take precedence over "self:read/self:write". If this is the case, it might be worth moving the `"user:read/user:write"` check above `"self:read/self:write"` and doing a return within the if blocks themselves so that the `"self:permission"` check isn't always required. The speed difference would be negligible but its nice to not have to do an unnecessary evaluation.

Reply:
Yes, good. I'm not sure why this was so hard but coming back to it now it's so obvious, thanks!